### PR TITLE
M5G-295: messaging input attachment FileInput (initial version for testing, not finished)

### PR DESCRIPTION
**Pull request by jonah-clever**
**Jira:**

https://clever.atlassian.net/browse/M5G-295

**Overview:**

Adding a `FileInput` to `MessagingInput` for uploading attachments. This is just a preliminary version for Portals team to be able to start testing. Clicking the `FileInput` opens a file chooser menu, but the store function does not do anything as of now.

**Screenshots/GIFs:**

Screen Shot 2021-05-19 at 2 59 12 PM

**Testing:**

- [ ] Unit tests
  -  I did not add new unit tests, but existing tests pass
- Manual tests:
  - [x] Chrome
  - [x] Safari
  - [ ] IE11
  - [x] Edge

**Roll Out:**

- Before merging:
  - [x] Updated docs
  - [x] Bumped version in `package.json`
    - backward-compatible component feature change? Run `npm version minor`
- After merging:
  - [ ] Deployed updated docs (`make deploy-docs`)
  - [ ] Posted in #eng if I made a breaking change to a beta component

### CODE REVIEW

**Comment by zohnnyclever** [comment]
oh nice! @dengjulie do you know how we might use this change to do a test in our mobile app?

File: docs/components/MessagingAttachmentView.jsx [hunk 1]
```diff
@@ -24,71 +24,6 @@ export default class MessagingAttachmentView extends React.PureComponent {
   static cssClass = cssClass;
```
Comment (jonah-clever) [inline docs/components/MessagingAttachmentView.jsx:24]: note: changes in this file are unrelated to the rest of the PR. Just some cleanup from recent changes to this component

File: docs/components/FileInputView.jsx [hunk 2]
```diff
@@ -170,6 +170,13 @@ export default class FileInputView extends React.PureComponent {
               optional: true,
               defaultValue: <code>FormElementSize.FULL_WIDTH</code>,
             },
+            {
+              name: "isMessagingAttachmentInput",
+              type: "boolean",
+              description:
+                "Optional boolean which dictates if this FileInput is being used for uploading message attachments. If so, the component is rendered specifically for that purpose",
+              optional: true,
+            },
```
Comment (jonjonray) [inline docs/components/FileInputView.jsx:179]: Wondering if we should avoid adding this to the docs for FileInput since we mostly use this prop internally from within MessagingInput

File: docs/components/TextAreaView.jsx [hunk 3]
```diff
@@ -277,6 +277,20 @@ export default class TextAreaView extends React.Component {
...
+              optional: true,
+            },
+            {
+              name: "storeAttachment",
+              type: "(file, callbacks) => void",
+              description:
+                "Optional store function to be used with the FileInput for attachment uploads. This is only used if showUploadAttachmentButton == true, and should always be defined in that case.",
+              optional: true,
+            },
           ]}
```
Comment (jonjonray) [inline docs/components/TextAreaView.jsx:294]: same comment as above about potentially not documenting this new prop since we're using it internally

File: src/FileInput/FileInput.less [hunk 4]
```diff
@@ -36,6 +36,26 @@
...
+  .FileInput--Icon {
+    margin: 0.25rem 0.5rem 0.25rem 0.75rem;
+    border-radius: 0.25rem;
+
+    &:hover {
+      fill: #f5f8fd;
+    }
+  }
+}
+
```
Comment (jonjonray) [inline src/FileInput/FileInput.less:58]: nit: might be worth using mixins for a few of these

height: @size_2xl
.padding--none()
.borderRadiusL()

File: src/FileInput/FileInput.tsx [hunk 5]
```diff
@@ -163,9 +184,11 @@ export class FileInput extends React.Component<Props, State> {
   }
 
   render() {
-    const { iconOnly, label } = this.props;
+    const { iconOnly, label, isMessagingAttachmentInput } = this.props;
     const dropzoneStyle = iconOnly ? { display: "inline-block" } : {};
 
+    // console.log(this.state);
```
Comment (jonjonray) [inline src/FileInput/FileInput.tsx:190]: Nit: console.log

File: src/FileInput/FileInput.tsx [hunk 6]
```diff
@@ -110,6 +130,7 @@ const propTypes = {
   store: PropTypes.func.isRequired,
   accept: PropTypes.string,
   size: PropTypes.string,
+  isMessagingAttachmentInput: PropTypes.bool,
```
Comment (jonjonray) [inline src/FileInput/FileInput.tsx:133]: Another thought here: wondering if there's an opportunity to potentially abstract this prop into something non-messaging related e.g. some kind of icon override since we're mainly checking this boolean so we can update the icon on line 203

File: src/MessagingInput/MessagingInput.tsx [hunk 7]
```diff
@@ -31,6 +31,10 @@ interface Props {
   labelText?: string;
   disableSendButton?: boolean;
   showReturnKeyInstructions?: boolean;
+  // TODO: is this needed or is the existance of MessagingInput.store sufficient to decide this?
```
Comment (jonjonray) [inline src/MessagingInput/MessagingInput.tsx:34]: I have no strong preference, can see an argument to be made for both cases (more explicit vs less props)

**Review by jonjonray — APPROVED** [review]
Left some comments but LGTM overall

File: src/MessagingInput/MessagingInput.tsx (hunk 7 above)
Comment (jonah-clever) [reply src/MessagingInput/MessagingInput.tsx:34]: I'm gonna go with more explicit for now, will remove comment

File: src/FileInput/FileInput.tsx [hunk 8]
```diff
@@ -204,6 +204,10 @@ export class FileInput extends React.Component<Props, State> {
                 "FileInput",
                 formElementSizeClassName(this.props.size),
                 this.props.className,
+                !!this.state.filename &&
+                  !this.state.success &&
+                  !this.state.error &&
+                  "FileInput--AttachmentIsUploading",
```
Comment (jonjonray) [inline src/FileInput/FileInput.tsx:210]: It seems like the styles are only getting applied from the MessagingInput `.less` file but we would still end up adding an extra class for every usage of `FileInput`, which might not be ideal. 

Was thinking it might make sense to add this class to the top level component (MessagingInput) and apply it from there. Something like

```
.MessagingInput--AttachmentIsUpload .FileInput { ... }
```

**Review by jonjonray — APPROVED** [review]
LGTM! Left a comment about the upload class name, but it's somewhat minor

File: src/FileInput/FileInput.tsx (hunk 8 above)
Comment (jonah-clever) [reply src/FileInput/FileInput.tsx:210]: I thought about this too. The issue is that the class is added here based on the state of the `FileInput`. I'm not sure exactly how to pass that info up to the `MessagingInput`, so the class could be set there. Open to ideas tho!

Comment (jonjonray) [reply src/FileInput/FileInput.tsx:210]: Ah yeah, that's a good point. The current approach makes sense then!
